### ISSUE TITLE
Move automatic method selection to C core

### ISSUE THREAD

**Issue by szhorvat** [igraph/igraph#2481]
There are several pieces of functionality where multiple methods can be used to calculate the same thing. In some cases, these is a single interface to all of them, e.g. `igraph_degree_sequence_game()`. In some other cases there are different functions, e.g. all the `igraph_distances_XXX()` functions.

I propose standardizing how we do this as follows, and starting some of this (simple) refactoring already for 1.0:

 - There will be a common interface, with a `method` parameter. This parameter will have an "automatic" setting, where the "best" method is chosen automatically. A big advantage of this is that it increases the code sharing between interfaces, which often already implement their automatic method selection. I think @krlmlr will be happy if `distances()` can finally be auto-generated for R (not possible with 0.10).
 - When different methods have their own specific settings, they should also have their own dedicated functions. For example, the VL and "edge switching" methods of `igraph_degree_sequence_game()` should have their own specific options to control e.g. the number of randomization steps. Such an option does not fit well into the common interface.

Example:

`igraph_distances()` would be renamed to `igraph_distances_unweighted()` and the name `igraph_distances()` would be recycled as follows:

```c
igraph_error_t igraph_distances(const igraph_t *graph,
                                const igraph_vector_t *weights,
                                const igraph_vs_t from,
                                const igraph_vs_t to,
                                igraph_neimode_t mode,
                                igraph_distances_method_t method);
```

There would be an `IGRAPH_DISTANCES_AUTOMATIC` method which selects the appropriate method in a similar way to what the high-level interfaces already do individually at the moment: Unweighted for `weights == NULL`, Dijkstra for positive weights, Johnson for negative weights and computing _most_ of the distance matrix, Bellman-Ford for negative weights and computing only a few rows, Floyd-Warshall for extremely dense graphs, etc.

**Comment by ntamas** [comment]
TODO list:

- [ ] `igraph_pagerank_algo_t` should be extended with an "automatic" member
- [ ] `igraph_barabasi_algorithm_t `should be extended with an "automatic" member
- [ ] `igraph_fas_algorithm_t` should be extended with an "automatic" member
- [ ] `igraph_fvs_algorithm_t` should be extended with an "automatic" member
- [ ] `igraph_degseq_t` should be extended with an "automatic" member.
- [ ] `igraph_random_tree_t` should be extended with an "automatic" member.
- [ ] `igraph_realize_degseq_t` should be extended with an "automatic" member.
- [ ] `IGRAPH_EIGEN_AUTO` needs to be aliased to `IGRAPH_EIGEN_AUTOMATIC` (or the other way round) because we already have `igraph_mst_algorithm_t` that uses "automatic" and not the shorter "auto" variant (and the same applies to `igraph_floyd_warshall_algorithm_t`).
- [ ] `igraph_distances_weighted()` should be added, _or_ `igraph_distances()` should be extended with a `weights` and an `algorithm` argument (although we can do away with the "algorithm" argument and just say that it always uses automatic selection).
- [ ] `igraph_get_shortest_paths_weighted()` should be added, _or_ `igraph_get_shortest_paths()` should be extended with a `weights` and optionally an `algorithm` argument.
- [ ] `igraph_get_all_shortest_paths_weighted()` should be added, _or_ `igraph_get_all_shortest_paths()` should be extended with a `weights` and optionally an `algorithm` argument.

(Personal note: I'm not a big fan of adding `algorithm` arguments everywhere; we can simply say that these function attempt to choose the best algorithm and if the user needs more precise control over the algorithm being selected, they should just call their algorithm of choice directly. In theory, any igraph function could have multiple possible implementations and it is not realistic to start adding an `algorithm` argument everywhere just because we _might_ add a new algorithm later).

Note that adding a new member to an existing enum is probably not ABI-breaking so we can do most of the items above any time, even after 1.0, except for the last three items.

**Comment by szhorvat** [comment]
> * `igraph_barabasi_algorithm_t `should be extended with an "automatic" member

I would leave this function alone. The "algorithms" do different things (multigraph vs simple graph). There is no default choice.

However, we might rename "algorithm" to something else to avoid this sort of confusion.

> * `igraph_fas_algorithm_t` should be extended with an "automatic" member
> * `igraph_fvs_algorithm_t` should be extended with an "automatic" member

I would leave these alone. IMO there is not a good default choice. Some methods are exact, some slow but give tight bounds, some fast but give large feedback sets. The tradeoff should be decided by the user, especially between exact vs approximate methods.

 

> * `igraph_degseq_t` should be extended with an "automatic" member.

This function is kind of a mess because some methods are more or less true alternatives, but others construct graphs of a different type.

The only two alternatives now are configuration_simple method vs edge_switching.  The rest all do different things: either don't sample uniformly or sample multigraphs instead of simple graphs (perhaps also not uniformly!) or sample connected graphs.

I'll think about whether we want to change the interface here, as this is part of the degseq project we're doing with Arnar and Zara, but it seems too complicated to reach a good decision within a month ... For this reason I would not block release on this.



> * `igraph_realize_degseq_t` should be extended with an "automatic" member.

Again these do different things, so I would not.

----

> and if the user needs more precise control over the algorithm being selected, they should just call their algorithm of choice directly.

You mean provide separate functions for each method/algorithm?  I'm fine with that because several algorithms will have their own parameters that can't be all crammed into a single function.

The main motivation for this request is to make the maintenance of high-level interfaces easier. The main feature I'd like to see is to add this for the shortest path / distances calculation. I would keep everything in a single function, i.e. extend `igraph_distances()` with a parameter instead of adding `igraph_distance_weighted()`. This is consistent with functions like `igraph_eccentricity()` and makes maintenance of high-level interfaces easier.

Let's keep in mind that if a function already has a method/algorithm parameter, then adding an "auto" option is not API-breaking, so we don't need to block 1.0 on these. The only one from the above list which fits this, and where I'd make the modification, is PageRank.

----

Do we want to standardize the naming of such parameters? `algo`? `algorithm`? `method`? I like `method`, but I'm fine with anything really.

**Comment by szhorvat** [comment]
**tl;dr** IMO the only task for 1.0 here is to add this feature to shortest-path related functions (last three in the TODO list above). I'd add a method parameter to the current unweighted functions instead of adding a new function.

**Comment by ntamas** [comment]
> Do we want to standardize the naming of such parameters? 

There's really no need to. We can invent as many "standards" as we want but even we are not able to remember them, let alone our users. IDEs have autocompletion, and parameter names don't really matter in C anyway, so why bother?

Okay, so then the plan is to:

 - [ ] Add a `weights` argument to `igraph_distances()`, opting for automatic selection of the algorithm depending on the weights vector
 - [ ] Add a `weights` argument to `igraph_get_shortest_paths()`, opting for automatic selection of the algorithm depending on the weights vector
 - [ ] Add a `weights` argument to `igraph_get_all_shortest_paths()`, opting for automatic selection of the algorithm depending on the weights vector

Are we on the same page?

**Comment by ntamas** [comment]
Fixed in 797f0a9.

There does not seem to be a consensus on the position of the `weights` argument in shortest path related functions; I added it to the second position now but there are a few other functions where it is somewhere back in the argument list. If that bothers you, feel free to send a PR to fix them.

**Comment by szhorvat** [comment]
@ntamas I still think it makes sense to add a `method`/`algorithm` parameter to the distances functions. This supports the high-level interfaces much better, as they can use a single function to expose this functionality. I do not support dropping method selection from high-level interfaces such as  R.

Do you object to my adding this parameter?

I do not mind removing the specific functions (`_dijkstra`, `_bellman_ford`, etc.) if you want to keep the API tidier. Most of these become redundant, unless they have extra parameters. I also do not mind keeping them.

CC @krlmlr @schochastics

**Comment by ntamas** [comment]
> Do you object to my adding this parameter?

No, feel free to go ahead.

The more specific functions could be removed if we are absolutely sure that these functions will not need additional parameters that are specific to them. Since they are well-established methods that have been around since like forever, it is probably okay to make them internal.
